packages/@headlessui-react/src/hooks/use-element-size.ts

```diff
@@ -12,15 +12,19 @@ export function useElementSize(
   unit = false
 ) {
   let element = ref === null ? null : 'current' in ref ? ref.current : ref
-  let [size, setSize] = useState(() => computeSize(element))
+  let [identity, forceRerender] = useReducer(() => ({}), {})
+
+  // When the element changes during a re-render, we want to make sure we
+  // compute the correct size as soon as possible. However, once the element is
+  // stable, we also want to watch for changes to the element. The `identity`
+  // state can be used to recompute the size.
+  let size = useMemo(() => computeSize(element), [element, identity])
```

Comment:
Now that `computeSize` is only used once, I was thinking about inlining it, but I kinda like the one-liner here. Thoughts?

```ts
  function computeSize(element: HTMLElement | null) {
    if (element === null) return { width: 0, height: 0 }
    let { width, height } = element.getBoundingClientRect()
    return { width, height }
  }
  
  // ---

  let size = useMemo(() => computeSize(element), [element, identity])
```

vs

```ts
  let size = useMemo(() => {
    if (element === null) return { width: 0, height: 0 }
    let { width, height } = element.getBoundingClientRect()
    return { width, height }
  }, [element, identity])
```

Reply:
Shouldn't we use `useLayoutEffect` for this instead given that it is guaranteed to fire before the DOM is flushed?

Reply:
Okay, seems fine to me then. 👍